eras/babbage/impl/src/Cardano/Ledger/Babbage/PParams.hs

```diff
@@ -0,0 +1,533 @@
+      _maxTxExUnits = fromSMaybe (_maxTxExUnits pp) (_maxTxExUnits ppup),
+      _maxBlockExUnits = fromSMaybe (_maxBlockExUnits pp) (_maxBlockExUnits ppup),
+      _maxValSize = fromSMaybe (_maxValSize pp) (_maxValSize ppup),
+      _collateralPercentage = fromSMaybe (_collateralPercentage pp) (_collateralPercentage ppup),
+      _maxCollateralInputs = fromSMaybe (_maxCollateralInputs pp) (_maxCollateralInputs ppup)
+    }
+
+-- ===================================================
+-- Figure 1: "Definitions Used in Protocol Parameters"
+
+-- The LangDepView is a key value pair. The key is the (canonically) encoded
+-- language tag and the value is the (canonically) encoded set of relevant
+-- protocol parameters
+data LangDepView = LangDepView {tag :: ByteString, params :: ByteString}
+  deriving (Eq, Show, Ord, Generic, NoThunks)
+
+getLanguageView ::
+  forall era.
+  PParams era ->
+  Language ->
+  LangDepView
+getLanguageView pp lang@PlutusV1 =
+  LangDepView -- The silly double bagging is to keep compatibility with a past bug
+    (serialize' (serialize' lang))
+    ( serialize'
+        ( serializeEncoding' $
+            maybe encodeNull enc $
+              Map.lookup lang (_costmdls pp)
+        )
+    )
+  where
+    enc (CostModel cm) = encodeFoldableAsIndefinite $ Map.elems cm
+getLanguageView pp lang@PlutusV2 =
+  LangDepView
+    (serialize' lang)
+    (serializeEncoding' $ maybe encodeNull toCBOR $ Map.lookup lang (_costmdls pp))
+
+encodeLangViews :: Set LangDepView -> Encoding
+encodeLangViews views = encodeMapLen n <> foldMap encPair ascending
+  where
+    n = fromIntegral (Set.size views) :: Word
+    ascending = sortBy (shortLex `on` tag) $ Set.toList views
+    encPair (LangDepView k v) = encodePreEncoded k <> encodePreEncoded v
+    shortLex :: ByteString -> ByteString -> Ordering
+    shortLex a b
+      | BS.length a < BS.length b = LT
+      | BS.length a > BS.length b = GT
+      | otherwise = compare a b
+
+-- | Turn an PParams' into a Shelley.Params'
+retractPP :: HKD f Coin -> HKD f UnitInterval -> HKD f Nonce -> PParams' f era2 -> Shelley.PParams' f era1
+retractPP
+  c
+  d
+  eE
+  (PParams ma mb mxBB mxT mxBH kd pd emx a n rho tau pv mnP _ _ _ _ _ _ _ _) =
+    Shelley.PParams ma mb mxBB mxT mxBH kd pd emx a n rho tau d eE pv c mnP
+
+-- | Given the missing pieces Turn a Shelley.PParams' into an Params'
+extendPP ::
+  Shelley.PParams' f era1 ->
+  HKD f Coin ->
+  HKD f (Map Language CostModel) ->
+  HKD f Prices ->
+  HKD f ExUnits ->
+  HKD f ExUnits ->
+  HKD f Natural ->
+  HKD f Natural ->
+  HKD f Natural ->
+  PParams' f era2
+extendPP
+  (Shelley.PParams ma mb mxBB mxT mxBH kd pd emx a n rho tau _d _eE pv _ mnP)
+  ada
+  cost
+  price
+  mxTx
+  mxBl
+  mxV
+  col
+  mxCol =
+    PParams ma mb mxBB mxT mxBH kd pd emx a n rho tau pv mnP ada cost price mxTx mxBl mxV col mxCol
+
+instance HasField "_d" (PParams era) UnitInterval where
+  getField _ = minBound
```

Comment:
is this sensible? it just required the least change to satisfy constraints on https://github.com/input-output-hk/cardano-ledger/tree/zachc/babbage-big. i chose `minBound` because that's what was chose in `emptyPParams`

Reply:
do you know what is requiring the `HasField` instance? The link just points to a branch. I suspect it is the reward calculation, it which case having this spurious instance might be the path of least resistance, saving us from coping a lot of reward calc code. and yea, minBound would be the best value.

Reply:
```
src/Cardano/Ledger/Babbage.hs:121:10: error:
    • Could not deduce (GHC.Records.HasField
                          "_d"
                          (PParams (BabbageEra c))
                          Cardano.Ledger.BaseTypes.UnitInterval)
        arising from the superclasses of an instance declaration
      from the context: API.ShelleyEraCrypto c
        bound by the instance declaration
        at src/Cardano/Ledger/Babbage.hs:121:10-64
    • In the instance declaration for 'API.ApplyBlock (BabbageEra c)'
    |
121 | instance API.ShelleyEraCrypto c => API.ApplyBlock (BabbageEra c)
    |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
some superclass constraint of `ApplyBlock`, i can't really figure out where this constraint arises.

Reply:
@TimSheard said he'd help investigate this. If it turns out it is needed (which is possible, the reward calc might be forcing it somehow) this this instance is good, but we should document why it is needed and the fact that it's a bit strange to have a HasField instance for a field it doesn't have...

Reply:
Where is the file Babbage.hs,  I can't find it . Makes it hard to reproduce the problem

Reply:
it's on the branch i linked above https://github.com/input-output-hk/cardano-ledger/tree/zachc/babbage-big

Reply:
@goolord I took a look at all the functions that need `_d`. It's going to be a lot of work to disentangle `_d`, and potentially add a lot of code duplication. I think what you've done, globally setting `d = 0` for Babbage, is the way to go. Let's document the `HasField "_d"` instance with a comment that there really is no such field, that we are just setting the value to zero since the reward calculation uses it. It is weird if I would rather set it to `0` than `minBound`? just seems more explicit...

Reply:
`UnitInterval` doesn't have its constructor exported and only provides `Bounded` and `BoundedRational` as afaict

Reply:
gotcha :+1: